### PR TITLE
Issue 522

### PR DESCRIPTION
**Pull request by balmas**
Summary of changes:

- adds support for versioning of settings, with the version number being included in the individual settings names that get saved to storage
- enables mid-session reset of settings, either back to defaults, or loaded from a new storage area (such as upon authentication)
- removed some unused code
- added a new storage adapters, RemoteAuthStorageAdapter which gets/saves settings to a remote service (which ended up being easier to implement as a new DynamoDB/Lambda service rather than trying to use the Auth0 metadata - see https://github.com/alpheios-project/user-settings-api)

### CODE REVIEW

File: src/lib/options/options.js [hunk 1]
```diff
@@ -10,32 +10,36 @@ export default class Options {
    * Mandatory fields:
    *    {string} domain - A domain name that defines options context
    *    {Object} items - An object that represents options that are exposed to the user. Each property is an option name.
-   * @param {Function<StorageAdapter>} StorageAdapter - A storage adapter implementation
+   * @param {StorageAdapter} storageAdapter - A storage adapter implementation
+   * @param {Object} adapterInitObj - an object which can be passed to the Storage Adapter to provide additional application state
```
Comment (kirlat) [inline src/lib/options/options.js]: Has this been implemented?

Comment (balmas) [reply src/lib/options/options.js]: oh, no that was from an earlier approach that I decided against. Thanks I will remove.

File: src/lib/options/remote-auth-storage-area.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * A proxy for the Alpheios user-settings-api
+   * It allows for storing key/value pairs to the api with an authorized user account
+   * @param {object} keysObject - An object containing one or more key/value pairs to be stored in storage.
+   * If a particular item already exists, its value will be updated.
+   * @return {Promise} - A promise that is resolved with with a void value if all key/value pairs are stored
+   * successfully. If at least on save operation fails, returns a rejected promise with an error information.
+   */
+  async set (keysObject) {
+    let [key, value] = Object.entries(keysObject)[0]
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/lib/options/remote-auth-storage-area.js]: well, this is the only way set it used right now, to set a single key/value pair. But you're right that the code here is inconsistent with the other 2 adapters, each of which account for the possibility of multiple being set at once. I was just being a bit lazy... will fix.

File: src/vue/components/panel-compact.vue [hunk 3]
```diff
@@ -264,9 +264,9 @@
            v-show="$store.getters['ui/isActiveTab']('options')"
            data-alpheios-ignore="all"
       >
-        <ui-settings></ui-settings>
-        <feature-settings></feature-settings>
-        <resource-settings></resource-settings>
+        <ui-settings :key="uiSettingsKey"></ui-settings>
+        <feature-settings :key="featureSettingsKey"></feature-settings>
+        <resource-settings :key="resourceSettingsKey"></resource-settings>
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/vue/components/panel-compact.vue]: they key is used only to force a re-render of the component. I was following the advice here https://michaelnthiessen.com/force-re-render

File: src/vue/components/panel-compact.vue [hunk 4]
```diff
@@ -444,6 +444,18 @@ export default {
       return classes
     },
 
+    uiSettingsKey() {
+      return `panel-compact-settings-ui-${this.$store.state.settings.uiResetCounter}`
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/vue/components/panel-compact.vue]: that's a good idea. will implement.

File: src/lib/controllers/ui-controller.js [hunk 5]
```diff
@@ -1292,6 +1342,22 @@ export default class UIController {
...
+    return this.resourceOptions
+  }
+
+  getUiOptions() {
+    return this.uiOptions
+  }
+
+  verboseMode() {
+    return this.uiOptions.items.verboseMode.currentValue === "verbose"
+  }
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/lib/controllers/ui-controller.js]: ok. I think it is reactive enough right now for this setting (e.g. if you change verboseMode you do immediately get the the status tab)

File: src/vue/components/font-size.vue [hunk 6]
```diff
@@ -3,20 +3,20 @@
     <label class="alpheios-setting__label">{{ l10n.getMsg('LABEL_RESKIN_SETTINGS') }}:</label>
     <div class="alpheios-setting__button-group alpheios-setting__control">
       <button
-          :class="{ active: activeButton === 12 }"
-          @click="changeFontSize(12)"
+          :class="{ active: activeButton === 'small' }"
+          @click="changeFontSize('small')"
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/vue/components/font-size.vue]: this whole bit of code confused me (See also #659 ) . I am ambivalent about which is better.

File: src/vue/components/panel-compact.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirlat) [reply src/vue/components/panel-compact.vue]: Oh, was not aware of such technique. Thanks for the info!

File: src/lib/options/remote-auth-storage-area.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    this.baseURL = auth.endpoints.settings,
+    this.requestContext = {
+      headers: {
+        common: {
+          Authorization: 'bearer ' + auth.accessToken,
+          'Content-Type': 'application/json'
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/lib/options/remote-auth-storage-area.js]: ok, good suggestion. will implement.

**Comment by balmas** [comment]
@kirlat can you take another look? I think I implemented all of your suggestions. Thanks!

**Pull request merged**